src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php

```diff
@@ -0,0 +1,433 @@
+
+        $this->controllerClass = $controllerClass;
+
+        return $this;
+    }
+
+    /**
+     * Creates the final ArrayCollection, returns it, and clears everything.
+     *
+     * @return RouteCollection
+     */
+    public function build()
+    {
+        $routeCollection = new RouteCollection();
+
+        foreach ($this->routes as $name => $route) {
+            if ($route instanceof Route) {
+                // auto-generate the route name if it's been marked
+                if ('_unnamed_route_' === substr($name, 0, 15)) {
+                    $name = $this->generateRouteName($route);
+                }
+
+                $this->ensureRouteController($route);
+
+                $route->setDefaults(array_merge($this->defaults, $route->getDefaults()));
+                $route->setRequirements(array_merge($this->requirements, $route->getRequirements()));
+                $route->setOptions(array_merge($this->options, $route->getOptions()));
+
+                if (null !== $this->prefix) {
+                    $route->setPath('/'.$this->prefix.$route->getPath());
+                }
+
+                if (!$route->getHost()) {
+                    $route->setHost($this->host);
+                }
+
+                if (!$route->getCondition()) {
+                    $route->setCondition($this->condition);
+                }
+
+                if (!$route->getSchemes()) {
+                    $route->setSchemes($this->schemes);
+                }
+
+                if (!$route->getMethods()) {
+                    $route->setMethods($this->methods);
+                }
+
+                $routeCollection->add($name, $route);
+            } else {
+                /* @var self $route */
+
+                $subCollection = $route->build();
+                $subCollection->addPrefix($this->prefix);
+
+                $routeCollection->addCollection($subCollection);
+            }
+        }
+
+        foreach ($this->resources as $resource) {
+            $routeCollection->addResource($resource);
+        }
+
+        return $routeCollection;
+    }
+
+    /**
+     * Attempts to safely prefix controllers with the controller class if necessary.
+     *
+     * @param Route $route
+     */
+    private function ensureRouteController(Route $route)
+    {
+        // only do work if there is a controller class set
+        if (!$this->controllerClass) {
+            return;
+        }
+
+        $controller = $route->getDefault('_controller');
+
+        // only apply controller class to a (non-empty) string
+        if (!is_string($controller) || !$controller) {
+            return;
+        }
+
+        // is the controller already a callable function/class?
+        if (method_exists($controller, '__invoke') || function_exists($controller)) {
+            return;
+        }
+
+        // is this already a controller format (a:b:c, or a:b, or a::b)?
+        if (false !== strpos($controller, ':')) {
+            return;
+        }
+
```

Comment:
You're absolutely right good catch. But there's no way to support that here unfortunately, and I think that's fine: if you're passing invokable service ids to your routes, you shouldn't set the controller class on your collection. That just doesn't make sense.

In a perfect world, I'd like to be able to detect this and throw an exception, but as this is an extra optional feature, I think if we can make this work for the 99% and allow the 1% to do their work just as easily, that's a win.